tensorflow_probability/python/distributions/empirical.py

```diff
@@ -0,0 +1,190 @@
+  (A multiset allows for multiple instances for each of its elements.)
+
+  #### Mathematical Details
+
+  The probability mass function (pmf) is,
+
+  ```none
+  pmf(k; A) = sum_i I(k)**[yi == k] / |A|
+  I(k)**[yi == k] == 1 if yi ∈ A and 0 otherwise.
+  ```
+
+  #### Examples
+
+  ```python
+
+  # Creates a empirical distribution with 3 observations.
+  dist = Empirical(samples=[0., 0., 1.])
+
+  # The length of first dimension of samples is number of samples.
+  dist.num_samples()
+
+  # Evaluate the pmf at 1, returning a scalar.
+  dist.prob(1.)
+
+  # Get 3 samples, returning a tensor with shape (3,).
+  dist.sample([3])
+  ```
+
+  """
+
+  def __init__(self,
+               samples,
+               validate_args=False,
+               allow_nan_stats=True,
+               name="Empirical"):
+    """Initialize an `Empirical` random variable.
+    Args:
+      samples: tf.Tensor representing a collection of samples. Its
+        outer (left-most) dimension determines the number of samples.
+      validate_args: Python `bool`, default `False`. When `True` distribution
+        parameters are checked for validity despite possibly degrading runtime
+        performance. When `False` invalid inputs may silently render incorrect
+        outputs.
+      allow_nan_stats: Python `bool`, default `True`. When `True`, statistics
+        (e.g., mean, mode, variance) use the value "`NaN`" to indicate the
+        result is undefined. When `False`, an exception is raised if one or
+        more of the statistic's batch members are undefined.
+      name: Python `str` name prefixed to Ops created by this class.
+
+    Raises:
+      ValueError: if `samples` is scalar.
+    """
+    parameters = locals()
+    with tf.name_scope(name, values=[samples]):
+      self._samples = tf.convert_to_tensor(samples, name="samples")
+      if (self._samples.shape.is_fully_defined() and
+          self._samples.shape == tf.TensorShape([])):
+        raise ValueError("samples must have >= 1 dimension, got []")
+      self._num_samples = tf.shape(self._samples)[0]
+
+    super(Empirical, self).__init__(
+        dtype=self._samples.dtype,
+        reparameterization_type=reparameterization.FULLY_REPARAMETERIZED,
+        validate_args=validate_args,
+        allow_nan_stats=allow_nan_stats,
+        parameters=parameters,
+        graph_parents=[self._samples, self._num_samples],
+        name=name)
+
+  @staticmethod
+  def _param_shapes(sample_shape):
+    return {"samples": tf.convert_to_tensor(sample_shape, dtype=tf.int32)}
+
+  @property
+  def samples(self):
+    """Distribution parameter."""
+    return self._samples
+
+  @property
+  def num_samples(self):
+    """Number of samples."""
+    return self._num_samples
+
+  def _batch_shape_tensor(self):
+    return tf.constant([], dtype=tf.int32)
+
+  def _batch_shape(self):
+    return tf.TensorShape([])
+
+  def _event_shape_tensor(self):
+    return tf.shape(self.samples)[1:]
+
+  def _event_shape(self):
```

Comment:
I'm probably just ignorant, but I'm not sure I understand what the problem is. AFAIK there's no issue with constructing, e.g., a `TransformedDistribution(Independent(whatever))`, even though Independent has varying ndims. What would go wrong with, e.g., TransformedDistribution(Empirical) under this implementation?

Currently this class doesn't support batch shape at all -- I was imagining that could be added later. We could do that backwards-compatibly via an init argument along the lines of `event_ndims=None`, where the default would preserve the existing behavior, but specifying a finite number of event ndims treats the remaining dims as batch dimensions. From there it'd be easy to build out VectorEmpirical, etc.

I'm not sure how much additional work it'd be to support batch shape. I think it'd be great, but don't want to block this contribution if it turns out to be nontrivial. @joeyhaohao have you thought at all about how to implement sampling and probs with batches? It'd be awesome if you want to do that in this PR, but if not I'd vote we move ahead and plan to iterate.

Reply:
Ah, I was actually under the impression that reinterpreted_batch_ndims was statically known, but looking at the code that isn't the case (and looking at TransformedDistribution shows that it can handle that), so I do agree I don't actually see an issue here.

In that case what you suggested can be done (i.e. instead of my _Empirical recommendation, you could just drop the underscore, and have the specific subclasses as mentioned).

For batches, I'm fine with not supporting them, but please add a note / TODO to add them (and please file a github issue with this).

Will take a deeper look at this PR, thanks for the discussion!

Reply:
Hi @davmre @srvasude , sorry for the late reply and thanks for the suggestions. I agree it's time to move forward to support batch shapes, and I will be working on it. 

@srvasude I think it makes more sense to treat input shape as [S, B1, B2, ..., Bk, E], where S is the sample dimension, B is the Batch dimension, and E for event dimension. Since it's fine to have input with varying dimension in this case, I am considering two ways to do this:
1. Add `event_ndims=None` in the `Empirical` parameters as @davmre suggested, where default would preserve scalar-like samples. I guess we don't have to implement VectorEmpirical or EmpiricalNd in this case as one can pass event_ndims=1 or whatever and we will treat the remaining dims are batch dimensions
2. Fixed shape for each class, like `Deterministic` and `VectorDeterministic`. (i.e. an `Empirical` for scalar-like samples, `VectorEmpirical` for vector-like samples). And I think we also need a private `_Empirical` to avoid duplicate codes, as the logic behind built-in methods `mean`, `sample`, `entropy` are similar.

I would vote for 1. It seems more concise and elegant for me. What do you think?

Reply:
Agreed on shape. In particular this is what we return from TF distributions, so doing something like Empirical(dist.sample(N)) looks quite nice.

1 seems good to me (having _Empirical is essentially the same as 1, except it is a hidden distribution).

@davmre: Do you have any ideas on how this would wrap up / be as an Edward2 Random Variable. In particular I'm curious how ed.Empirical would work (would it be as simple as ed.Empirical(ed.OtherRV(), but with ed.Empirical inferring what the event_ndims are from ed.OtherRV ?)